### ISSUE TITLE
Inline mappings

### ISSUE THREAD

**Issue by siberianguy** [agileobjects/AgileMapper#25]
We have [this conversation](https://github.com/AutoMapper/AutoMapper/issues/2317) about inline mappings in AutoMapper. I'm considering changing the library if AutoMapper doesn't allow to do that, so I'm curious what is your stance on this?

**Comment by SteveWilkes** [comment]
Hi!

AgileMapper doesn't *require* a '`CreateMap`' configuration line, so the most basic inline case is just:

```C#
var course = Mapper.Map(message).ToANew<Course>();
```

The mapping function is created the first time the mapper is given a `Message` to map to a `Course`, and reused after that.

For cases requiring configuration, its my intention to enable passing the configuration at the time the mapping is performed:

```C#
var course = Mapper
    .Map(message)
    .ToANew<Course>(config => config
        .Map((m, c) => m.MessageCourseId)
        .To(c => c.CourseId));
```

For a small performance hit in looking up the cached mapping function, this puts custom mapping logic where I'd intuit it would be and where I'd say it makes the most sense - where the mapping is actually *performed*.

If the same configuration needs to specified more than once, it could be done using a method:

```C#
var course = Mapper
    .Map(message)
    .ToANew<Course>(WithMessageToCourseConfig);

var course2 = Mapper
    .Map(message2)
    .ToANew<Course>(WithMessageToCourseConfig);

private void WithMessageToCourseConfig(MapperConfigObj config)
{
    config
        .Map((m, c) => m.MessageCourseId)
        .To(c => c.CourseId)
}
```

...something like that.

This is probably a couple of months down the line, but a feature I'm heading towards.

**Comment by siberianguy** [comment]
Thanks for your answer! I totally agree with your logic. I don't know why AutoMapper went the way it did but the approach you described makes much more sense to me.

Quite off topic. ToANew method name seems a little off to me. Did you consider ToNew (we don't have to be 100% grammatically correct for names) or even better just To?

**Comment by SteveWilkes** [comment]
Hmmm, of the options:

```C#
Mapper.Map(message).To<Course>();
Mapper.Map(message).ToNew<Course>();
Mapper.Map(message).ToANew<Course>();
```

...I prefer `.ToANew()`, as I like APIs to read as grammatically-correct sentences where reasonable. Wouldn't be difficult for anyone to add an extension method to enable `.To()` or `.ToNew()` if that was their preference, though.

Thanks for the feedback :)

**Comment by siberianguy** [comment]
I just don't remember any serious API in .NET that would use A in a similar way. So it seems like you're going against the naming convention. It's okay in your own projects (I do it all the time), might be not the best idea for a public library. I feel like a better approach would be to implement a widely accepted To as the default option and keep your preferences as an extension method (so inversion of what we have now).

**Comment by SteveWilkes** [comment]
I certainly don't disagree strongly with you, I just have a couple of reservations:

`To` is used in the configuration API to apply a configured item to creation, merge and update mappings:

```C#
Mapper.WhenMapping
    .From<Message>()
    .Over<Course>() // Apply only to update mappings
    .Map((m, c) => m.MessageCourseId)
    .To(c => c.CourseId)

Mapper.WhenMapping
    .From<Message>()
    .ToANew<Course>() // Apply only to creation mappings
    .Map((m, c) => m.MessageCourseId)
    .To(c => c.CourseId)

Mapper.WhenMapping
    .From<Message>()
    .To<Course>() // Apply to updates, creations and merges
    .Map((m, c) => m.MessageCourseId)
    .To(c => c.CourseId)
```

So `ToANew` keeps the APIs consistent.

And while I certainly agree it's clear what a `To` method would do, for me `ToANew` is more explicit and reads better - it removes the (admittedly tiny) cognitive move required to get from 'to the `Course` type' to 'to a new instance of the `Course` type'. That's probably the reason why we both use 'A' in our own project APIs, and I think it justifies the 4 extra characters :)

**Comment by hcesar** [comment]
ToANew looks very strange to me.

Just **To** should be enough.

**Comment by siberianguy** [comment]
@hcesar Agree

**Comment by SteveWilkes** [comment]
Fair enough, re: 'To' - I'll solicit some more feedback and go from there :)

Incidentally, I'm now adding inline configuration on the [InlineConfiguration](/agileobjects/AgileMapper/tree/InlineConfiguration) branch.

**Comment by siberianguy** [comment]
@SteveWilkes Interesting. What's your current syntax choice for InlineConfiguration? You might be interesting to checkout the AutoMapper thread I mentioned in the beginning and so see how this feature is being designed there.

**Comment by SteveWilkes** [comment]
Syntax so far is exactly as I wrote in my first comment on this thread. I've pushed a couple of [passing tests](/agileobjects/AgileMapper/blob/InlineConfiguration/AgileMapper.UnitTests/Configuration/Inline/WhenConfiguringMappingInline.cs) which have examples. I haven't measured the performance overhead in looking up a compatible mapper with which to perform an 'inline' mapping, but it's going to be pretty tiny.

As you mentioned in the AutoMapper thread, if your supplied mapper configuration is invalid (defining two different data sources for the same target member, for example), you'll get a MappingConfigurationException when you try to perform the mapping.

**Comment by SteveWilkes** [comment]
This has now been implemented in [v0.15](https://nuget.org/packages/AgileObjects.AgileMapper) - I've updated the Wiki and added a [dedicated page](/agileobjects/AgileMapper/wiki/Inline-Configuration) for inline configuration.

I'm still soliciting feedback on ToANew being renamed, and will revisit it later.

Thanks for the feedback! :)
